{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "LibraryError",
     "evalue": "Edges must be listed in (time[parent], child, left) order; time[parent] order violated",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m--------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mLibraryError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-df0debb00d1d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    104\u001b[0m     \u001b[0mlength\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnbp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m     \u001b[0mrecombination_rate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrho\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m     mutation_rate=mu)#,\n\u001b[0m\u001b[1;32m    107\u001b[0m     \u001b[0;31m#random_seed=seed)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m \u001b[0;31m#print((treeseq.num_trees, treeseq.num_sites))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/storage/work/jzh774/miniconda3/lib/python3.7/site-packages/msprime/simulations.py\u001b[0m in \u001b[0;36msimulate\u001b[0;34m(sample_size, Ne, length, recombination_rate, recombination_map, mutation_rate, population_configurations, migration_matrix, demographic_events, samples, model, record_migrations, random_seed, mutation_generator, num_replicates, from_ts, start_time, end_time, record_full_arg, num_labels)\u001b[0m\n\u001b[1;32m    486\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mnum_replicates\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    487\u001b[0m         return next(_replicate_generator(\n\u001b[0;32m--> 488\u001b[0;31m             sim, mutation_generator, 1, provenance_dict, end_time))\n\u001b[0m\u001b[1;32m    489\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    490\u001b[0m         return _replicate_generator(\n",
      "\u001b[0;32m/storage/work/jzh774/miniconda3/lib/python3.7/site-packages/msprime/simulations.py\u001b[0m in \u001b[0;36m_replicate_generator\u001b[0;34m(sim, mutation_generator, num_replicates, provenance_dict, end_time)\u001b[0m\n\u001b[1;32m    163\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum_replicates\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m         \u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mend_time\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 165\u001b[0;31m         \u001b[0mtree_sequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_tree_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutation_generator\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprovenance_record\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    166\u001b[0m         \u001b[0;32myield\u001b[0m \u001b[0mtree_sequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    167\u001b[0m         \u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/storage/work/jzh774/miniconda3/lib/python3.7/site-packages/msprime/simulations.py\u001b[0m in \u001b[0;36mget_tree_sequence\u001b[0;34m(self, mutation_generator, provenance_record)\u001b[0m\n\u001b[1;32m    723\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mpop_config\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpopulation_configurations\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    724\u001b[0m                 \u001b[0mtables\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpopulations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_row\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetadata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpop_config\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoded_metadata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 725\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mtables\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree_sequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    726\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    727\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mreset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/storage/work/jzh774/miniconda3/lib/python3.7/site-packages/tskit/tables.py\u001b[0m in \u001b[0;36mtree_sequence\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1534\u001b[0m         \u001b[0;34m:\u001b[0m\u001b[0mrtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mTreeSequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1535\u001b[0m         \"\"\"\n\u001b[0;32m-> 1536\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mtskit\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTreeSequence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_tables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1537\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1538\u001b[0m     def simplify(\n",
      "\u001b[0;32m/storage/work/jzh774/miniconda3/lib/python3.7/site-packages/tskit/trees.py\u001b[0m in \u001b[0;36mload_tables\u001b[0;34m(cls, tables)\u001b[0m\n\u001b[1;32m   2175\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload_tables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2176\u001b[0m         \u001b[0mts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_tskit\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTreeSequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2177\u001b[0;31m         \u001b[0mts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_tables\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtables\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mll_tables\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2178\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mTreeSequence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2179\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mLibraryError\u001b[0m: Edges must be listed in (time[parent], child, left) order; time[parent] order violated"
     ]
    }
   ],
   "source": [
    "# simulate 10 demes of African American populations\n",
    "# African American: European with African admixture 16 generations ago\n",
    "# Note, this version deleted EUR-ASN split and all ASN-related migration event\n",
    "## in order to run this file, the random seed number is needed to add after script\n",
    "### Note, this version simulate 10Mb chromosome length\n",
    "import msprime, sys\n",
    "from math import log\n",
    "from math import exp\n",
    "#seed = int(sys.argv[1]) # input the seed\n",
    "#seed = 2\n",
    "\n",
    "mu=1.25e-8 # mutation rate per bp\n",
    "rho=1e-8 # recombination rate per bp\n",
    "nbp=1e8 # generate 100Mb\n",
    "#nbp=1e7 # generate 10Mb\n",
    "N0=7310 # initial population size\n",
    "Thum=5920 # time (gens) of advent of modern humans\n",
    "Naf=14474 # size of african population\n",
    "Tooa=2040 # number of generations back to Out of Africa\n",
    "Nb=1861 # size of out of africa population\n",
    "mafb=1.5e-4 # migration rate Africa and Out-of-Afica\n",
    "Teu=920 # number generations back to Asia-Euroupe split\n",
    "Neu=1032; Nas=554 # bottleneck population sizes\n",
    "mafeu=2.5e-5;mafas=7.8e-6;meuas=3.11e-5 #mig.rates\n",
    "reu=0.0038 #growth rate per generation in Europe\n",
    "ras=0.0048 #growth rate per generation in Asia\n",
    "Tadmix=16 # time of admixture\n",
    "Nadmix=30000 # initial size of admixed population\n",
    "radmix=.05 # growth rate of admixed population\n",
    "mdeme=0.05/2 # migration rate between neighboring demes\n",
    "\n",
    "# pop0 is Africa, pop1 is Europe,  pop2-11 is admixed\n",
    "refsamplesize=200\n",
    "admsamplesize=1000 #sample size of admixted pop\n",
    "pop_config = [\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Naf, growth_rate=0.0),\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Neu*exp(reu*Teu), growth_rate=reu),\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm1\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm2\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm3\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm4\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm5\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm6\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm7\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm8\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm9\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix) #adm10   \n",
    "]\n",
    "mig_mat=[\n",
    "    [0,mafeu,0,0,0,0,0,0,0,0,0,0],\n",
    "    [mafeu,0,0,0,0,0,0,0,0,0,0,0],\n",
    "    [0,0,0,mdeme,0,0,0,0,0,0,0,0],\n",
    "    [0,0,mdeme,0,mdeme,0,0,0,0,0,0,0],\n",
    "    [0,0,0,mdeme,0,mdeme,0,0,0,0,0,0],\n",
    "    [0,0,0,0,mdeme,0,mdeme,0,0,0,0,0],\n",
    "    [0,0,0,0,0,mdeme,0,mdeme,0,0,0,0],\n",
    "    [0,0,0,0,0,0,mdeme,0,mdeme,0,0,0],\n",
    "    [0,0,0,0,0,0,0,mdeme,0,mdeme,0,0],\n",
    "    [0,0,0,0,0,0,0,0,mdeme,0,mdeme,0],\n",
    "    [0,0,0,0,0,0,0,0,0,mdeme,0,mdeme],\n",
    "    [0,0,0,0,0,0,0,0,0,0,mdeme,0]\n",
    "]\n",
    "\n",
    "# Admixture event, 0.8 Africa, 0.2 Europe in total\n",
    "# 0 Afr, 1 Eur, source 2-11 admixed \n",
    "admixture_event = [msprime.MassMigration(time=Tadmix, source=2, destination=0, proportion=0.5), # Afr 0.5\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=2, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=3, destination=0, proportion=0.65), # Afr 0.65\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=3, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=4, destination=0, proportion=0.75), # Afr 0.75\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=4, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=5, destination=0, proportion=0.78), # Afr 0.78\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=5, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=6, destination=0, proportion=0.5), # Afr 0.8\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=6, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=7, destination=0, proportion=0.82), # Afr 0.82\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=7, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=8, destination=0, proportion=0.85), # Afr 0.85\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=8, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=9, destination=0, proportion=0.9), # Afr 0.9\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=9, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=10, destination=0, proportion=0.95), # Afr 0.95\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=10, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=11, destination=0, proportion=1.0), # Afr 1\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=11, destination=1, proportion=1.0) # Eur\n",
    "                  ] \n",
    "\n",
    "# Out of Africa event\n",
    "ooa_event = [\n",
    "    msprime.MigrationRateChange(time=Tooa, rate=0.0),\n",
    "    msprime.MassMigration(time=Tooa+0.0001, source=1, destination=0, proportion=1.0)\n",
    "]\n",
    "\n",
    "# initial population size\n",
    "init_event=[msprime.PopulationParametersChange(time=Thum, initial_size=N0, population_id=0)]\n",
    "\n",
    "#events = admixture_event + eu_event + ooa_event + init_event\n",
    "events = admixture_event + ooa_event + init_event\n",
    "\n",
    "treeseq = msprime.simulate(\n",
    "    population_configurations=pop_config,\n",
    "    migration_matrix=mig_mat,\n",
    "    demographic_events=events,\n",
    "    length=nbp,\n",
    "    recombination_rate=rho, \n",
    "    mutation_rate=mu)#,\n",
    "    #random_seed=seed)\n",
    "#print((treeseq.num_trees, treeseq.num_sites))\n",
    "\n",
    "# output genotype data\n",
    "# deal with Plink error \"Sample ID ends with \"_0\", which induces an invalid IID of '0'.\" # code comes from https://tskit.readthedocs.io/en/latest/_modules/tskit/trees.html\n",
    "# diploid\n",
    "filename=f'AA_10Mb_10deme.vcf'\n",
    "n_dip_indv = int(treeseq.num_samples / 2)\n",
    "indv_names = [f\"AA_{str(i)}indv\" for i in range(n_dip_indv)]\n",
    "with open(filename, \"w\") as vcf_file:\n",
    "    treeseq.write_vcf(vcf_file, ploidy=2, individual_names=indv_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
