{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulate 10 demes of African American populations, with equal admix pop size\n",
    "# African American: European with African admixture 16 generations ago\n",
    "# Note, this version deleted EUR-ASN split and all ASN-related migration event\n",
    "## in order to run this file, the random seed number is needed to add after script\n",
    "### Note, this version simulate 10Mb chromosome length\n",
    "import msprime, sys\n",
    "from math import log\n",
    "from math import exp\n",
    "#seed = int(sys.argv[1]) # input the seed\n",
    "seed = 1\n",
    "\n",
    "mu=1.25e-8 # mutation rate per bp\n",
    "rho=1e-8 # recombination rate per bp\n",
    "#nbp=1e8 # generate 100Mb\n",
    "nbp=1e7 # generate 10Mb\n",
    "N0=7310 # initial population size\n",
    "Thum=5920 # time (gens) of advent of modern humans\n",
    "Naf=14474 # size of african population\n",
    "Tooa=2040 # number of generations back to Out of Africa\n",
    "Nb=1861 # size of out of africa population\n",
    "mafb=1.5e-4 # migration rate Africa and Out-of-Afica\n",
    "Teu=920 # number generations back to Asia-Euroupe split\n",
    "Neu=1032; Nas=554 # bottleneck population sizes\n",
    "mafeu=2.5e-5;mafas=7.8e-6;meuas=3.11e-5 #mig.rates\n",
    "reu=0.0038 #growth rate per generation in Europe\n",
    "ras=0.0048 #growth rate per generation in Asia\n",
    "Tadmix=16 # time of admixture\n",
    "Nadmix=30000 # initial size of admixed population\n",
    "radmix=.05 # growth rate of admixed population\n",
    "mdeme=0.05/2 # migration rate between neighboring demes\n",
    "\n",
    "# pop0 is Africa, pop1 is Europe,  pop2-11 is admixed\n",
    "refsamplesize=200\n",
    "admsamplesize=100 #sample size of admixted pop\n",
    "pop_config = [\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Naf, growth_rate=0.0),\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Neu*exp(reu*Teu), growth_rate=reu),\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm1\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm2\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm3\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm4\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm5\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm6\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm7\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm8\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix), #adm9\n",
    "    msprime.PopulationConfiguration(sample_size=admsamplesize, initial_size=Nadmix*exp(radmix*Tadmix), growth_rate=radmix) #adm10   \n",
    "]\n",
    "mig_mat=[\n",
    "    [0,mafeu,0,0,0,0,0,0,0,0,0,0],\n",
    "    [mafeu,0,0,0,0,0,0,0,0,0,0,0],\n",
    "    [0,0,0,mdeme,0,0,0,0,0,0,0,0],\n",
    "    [0,0,mdeme,0,mdeme,0,0,0,0,0,0,0],\n",
    "    [0,0,0,mdeme,0,mdeme,0,0,0,0,0,0],\n",
    "    [0,0,0,0,mdeme,0,mdeme,0,0,0,0,0],\n",
    "    [0,0,0,0,0,mdeme,0,mdeme,0,0,0,0],\n",
    "    [0,0,0,0,0,0,mdeme,0,mdeme,0,0,0],\n",
    "    [0,0,0,0,0,0,0,mdeme,0,mdeme,0,0],\n",
    "    [0,0,0,0,0,0,0,0,mdeme,0,mdeme,0],\n",
    "    [0,0,0,0,0,0,0,0,0,mdeme,0,mdeme],\n",
    "    [0,0,0,0,0,0,0,0,0,0,mdeme,0]\n",
    "]\n",
    "\n",
    "# Admixture event, 0.8 Africa, 0.2 Europe in total\n",
    "# 0 Afr, 1 Eur, source 2-11 admixed \n",
    "admixture_event = [msprime.MassMigration(time=Tadmix, source=2, destination=0, proportion=0.5), # Afr 0.5\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=2, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=3, destination=0, proportion=0.65), # Afr 0.65\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=3, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=4, destination=0, proportion=0.75), # Afr 0.75\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=4, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=5, destination=0, proportion=0.78), # Afr 0.78\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=5, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=6, destination=0, proportion=0.5), # Afr 0.8\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=6, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=7, destination=0, proportion=0.82), # Afr 0.82\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=7, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=8, destination=0, proportion=0.85), # Afr 0.85\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=8, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=9, destination=0, proportion=0.9), # Afr 0.9\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=9, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=10, destination=0, proportion=0.95), # Afr 0.95\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=10, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=11, destination=0, proportion=1.0), # Afr 1\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=11, destination=1, proportion=1.0) # Eur\n",
    "                  ] \n",
    "\n",
    "# Out of Africa event\n",
    "ooa_event = [\n",
    "    msprime.MigrationRateChange(time=Tooa, rate=0.0),\n",
    "    msprime.MassMigration(time=Tooa+0.0001, source=1, destination=0, proportion=1.0)\n",
    "]\n",
    "\n",
    "# initial population size\n",
    "init_event=[msprime.PopulationParametersChange(time=Thum, initial_size=N0, population_id=0)]\n",
    "\n",
    "#events = admixture_event + eu_event + ooa_event + init_event\n",
    "events = admixture_event + ooa_event + init_event\n",
    "\n",
    "treeseq = msprime.simulate(\n",
    "    population_configurations=pop_config,\n",
    "    migration_matrix=mig_mat,\n",
    "    demographic_events=events,\n",
    "    length=nbp,\n",
    "    recombination_rate=rho, \n",
    "    mutation_rate=mu,\n",
    "    random_seed=seed)\n",
    "#print((treeseq.num_trees, treeseq.num_sites))\n",
    "\n",
    "# output genotype data\n",
    "# deal with Plink error \"Sample ID ends with \"_0\", which induces an invalid IID of '0'.\" # code comes from https://tskit.readthedocs.io/en/latest/_modules/tskit/trees.html\n",
    "# diploid\n",
    "filename=f'AA_10Mb_10deme.vcf'\n",
    "n_dip_indv = int(treeseq.num_samples / 2)\n",
    "indv_names = [f\"AA_{str(i)}indv\" for i in range(n_dip_indv)]\n",
    "with open(filename, \"w\") as vcf_file:\n",
    "    treeseq.write_vcf(vcf_file, ploidy=2, individual_names=indv_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulate 11 demes of African American populations, with not equal adm pop size, but equal anc interval\n",
    "# African American: European with African admixture 16 generations ago\n",
    "# Note, this version deleted EUR-ASN split and all ASN-related migration event\n",
    "## in order to run this file, the random seed number is needed to add after script\n",
    "### Note, this version simulate 10Mb chromosome length\n",
    "import msprime, sys\n",
    "from math import log\n",
    "from math import exp\n",
    "#seed = int(sys.argv[1]) # input the seed\n",
    "#seed = 1\n",
    "\n",
    "mu=1.25e-8 # mutation rate per bp\n",
    "rho=1e-8 # recombination rate per bp\n",
    "nbp=1e8 # generate 100Mb\n",
    "#nbp=1e7 # generate 10Mb\n",
    "N0=7310 # initial population size\n",
    "Thum=5920 # time (gens) of advent of modern humans\n",
    "Naf=14474 # size of african population\n",
    "Tooa=2040 # number of generations back to Out of Africa\n",
    "Nb=1861 # size of out of africa population\n",
    "mafb=1.5e-4 # migration rate Africa and Out-of-Afica\n",
    "Teu=920 # number generations back to Asia-Euroupe split\n",
    "Neu=1032; Nas=554 # bottleneck population sizes\n",
    "mafeu=2.5e-5;mafas=7.8e-6;meuas=3.11e-5 #mig.rates\n",
    "reu=0.0038 #growth rate per generation in Europe\n",
    "ras=0.0048 #growth rate per generation in Asia\n",
    "Tadmix=16 # time of admixture\n",
    "Nadmix=30000 # initial size of admixed population\n",
    "radmix=.05 # growth rate of admixed population\n",
    "mdeme=0.05/2 # migration rate between neighboring demes\n",
    "\n",
    "# pop0 is Africa, pop1 is Europe,  pop2-11 is admixed\n",
    "refsamplesize=200\n",
    "#admsamplesize=1000 #sample size of admixted pop\n",
    "# sample size of each admixed pop\n",
    "adm1=30\n",
    "adm2=0\n",
    "adm3=0\n",
    "adm4=45\n",
    "adm5=15\n",
    "adm6=45\n",
    "adm7=152\n",
    "adm8=500\n",
    "adm9=622\n",
    "adm10=561\n",
    "adm11=30\n",
    "pop_config = [\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Naf, growth_rate=0.0), #AFR\n",
    "    msprime.PopulationConfiguration(sample_size=refsamplesize, initial_size=Neu*exp(reu*Teu), growth_rate=reu), #EUR\n",
    "    msprime.PopulationConfiguration(sample_size=adm1, initial_size=Nadmix*exp(radmix*Tadmix)), #adm1\n",
    "    msprime.PopulationConfiguration(sample_size=adm2, initial_size=Nadmix*exp(radmix*Tadmix)), #adm2\n",
    "    msprime.PopulationConfiguration(sample_size=adm3, initial_size=Nadmix*exp(radmix*Tadmix)), #adm3\n",
    "    msprime.PopulationConfiguration(sample_size=adm4, initial_size=Nadmix*exp(radmix*Tadmix)), #adm4\n",
    "    msprime.PopulationConfiguration(sample_size=adm5, initial_size=Nadmix*exp(radmix*Tadmix)), #adm5\n",
    "    msprime.PopulationConfiguration(sample_size=adm6, initial_size=Nadmix*exp(radmix*Tadmix)), #adm6\n",
    "    msprime.PopulationConfiguration(sample_size=adm7, initial_size=Nadmix*exp(radmix*Tadmix)), #adm7\n",
    "    msprime.PopulationConfiguration(sample_size=adm8, initial_size=Nadmix*exp(radmix*Tadmix)), #adm8\n",
    "    msprime.PopulationConfiguration(sample_size=adm9, initial_size=Nadmix*exp(radmix*Tadmix)), #adm9\n",
    "    msprime.PopulationConfiguration(sample_size=adm10, initial_size=Nadmix*exp(radmix*Tadmix)), #adm10\n",
    "    msprime.PopulationConfiguration(sample_size=adm11, initial_size=Nadmix*exp(radmix*Tadmix)) #adm11\n",
    "    \n",
    "]\n",
    "mig_mat=[\n",
    "    [    0,mafeu,0,0,0,0,0,0,0,0,0,0,0], #AFR\n",
    "    [mafeu,0,    0,0,0,0,0,0,0,0,0,0,0], #EUR\n",
    "    [    0,0,0,mdeme,0,0,0,0,0,0,0,0,0], #ADM1\n",
    "    [0,0,mdeme,0,mdeme,0,0,0,0,0,0,0,0], #ADM2\n",
    "    [0,0,0,mdeme,0,mdeme,0,0,0,0,0,0,0], #ADM3\n",
    "    [0,0,0,0,mdeme,0,mdeme,0,0,0,0,0,0], #ADM4\n",
    "    [0,0,0,0,0,mdeme,0,mdeme,0,0,0,0,0], #ADM5\n",
    "    [0,0,0,0,0,0,mdeme,0,mdeme,0,0,0,0], #ADM6\n",
    "    [0,0,0,0,0,0,0,mdeme,0,mdeme,0,0,0], #ADM7\n",
    "    [0,0,0,0,0,0,0,0,mdeme,0,mdeme,0,0], #ADM8\n",
    "    [0,0,0,0,0,0,0,0,0,mdeme,0,mdeme,0], #ADM9\n",
    "    [0,0,0,0,0,0,0,0,0,0,mdeme,0,mdeme], #ADM10\n",
    "    [0,0,0,0,0,0,0,0,0,0,0,mdeme,0]  #ADM11\n",
    "]\n",
    "\n",
    "# Admixture event, 0.8 Africa, 0.2 Europe in total\n",
    "# 0 Afr, 1 Eur, source 2-12 admixed \n",
    "admixture_event = [msprime.MassMigration(time=Tadmix, source=2, destination=0, proportion=0.0), # Afr 0.0\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=2, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=3, destination=0, proportion=0.1), # Afr 0.1\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=3, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=4, destination=0, proportion=0.2), # Afr 0.2\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=4, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=5, destination=0, proportion=0.3), # Afr 0.3\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=5, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=6, destination=0, proportion=0.4), # Afr 0.4\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=6, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=7, destination=0, proportion=0.5), # Afr 0.5\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=7, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=8, destination=0, proportion=0.6), # Afr 0.6\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=8, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=9, destination=0, proportion=0.7), # Afr 0.7\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=9, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=10, destination=0, proportion=0.8), # Afr 0.8\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=10, destination=1, proportion=1.0), # Eur\n",
    "                   msprime.MassMigration(time=Tadmix, source=11, destination=0, proportion=0.9), # Afr 0.9\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=11, destination=1, proportion=1.0), # Eur                   \n",
    "                   msprime.MassMigration(time=Tadmix, source=12, destination=0, proportion=1.0), # Afr 1\n",
    "                   msprime.MassMigration(time=Tadmix+0.0001, source=12, destination=1, proportion=1.0) # Eur\n",
    "                  ] \n",
    "\n",
    "# Out of Africa event\n",
    "ooa_event = [\n",
    "    msprime.MigrationRateChange(time=Tooa, rate=0.0),\n",
    "    msprime.MassMigration(time=Tooa+0.0001, source=1, destination=0, proportion=1.0)\n",
    "]\n",
    "\n",
    "# initial population size\n",
    "init_event=[msprime.PopulationParametersChange(time=Thum, initial_size=N0, population_id=0)]\n",
    "\n",
    "#events = admixture_event + eu_event + ooa_event + init_event\n",
    "events = admixture_event + ooa_event + init_event\n",
    "\n",
    "treeseq = msprime.simulate(\n",
    "    population_configurations=pop_config,\n",
    "    migration_matrix=mig_mat,\n",
    "    demographic_events=events,\n",
    "    length=nbp,\n",
    "    recombination_rate=rho, \n",
    "    mutation_rate=mu)#,\n",
    "    #random_seed=seed)\n",
    "#print((treeseq.num_trees, treeseq.num_sites))\n",
    "\n",
    "# output genotype data\n",
    "# deal with Plink error \"Sample ID ends with \"_0\", which induces an invalid IID of '0'.\" # code comes from https://tskit.readthedocs.io/en/latest/_modules/tskit/trees.html\n",
    "# diploid\n",
    "filename=f'AA_100Mb_11deme.vcf'\n",
    "n_dip_indv = int(treeseq.num_samples / 2)\n",
    "indv_names = [f\"AA_{str(i)}indv\" for i in range(n_dip_indv)]\n",
    "with open(filename, \"w\") as vcf_file:\n",
    "    treeseq.write_vcf(vcf_file, ploidy=2, individual_names=indv_names)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
