---
title: "simple_simulation_addenv"
author: "Jinguo Huang"
date: "9/22/2021"
output: html_document
---

Change the simple_simulation to make Vtotal=1
```{r}
library(data.table)
library(ggplot2)
library(dplyr)
fbar = 0.5 # frequency in the parental population
theta = 0.4 #admixture fraction
nloci = 1000 #number of loci, not necessary causal variants
vgt = 1 #desired total genetic variance
vgb = 0.3 #desired genetic variance in between
#vgb = vgt*Fst
fst = vgb/vgt
#next, calculate frequency difference required to generate that Fst
#fst = (f1 - f2)^2 / 4*pbar*(1-pbar)
fdiff = sqrt(fst*4*fbar*(1-fbar))
#solve system of linear equation
f1_f2 = solve(a = matrix(c(0.5, 1, 0.5, -1), nrow = 2, ncol = 2), 
              b = c(fbar, fdiff))
# 0.5*f1+0.5*f2=fbar
# f1-f2=fdiff
# given fbar and fdiff, solve the linear equation to get f1,f2
f1 = f1_f2[1]
f2 = f1_f2[2]
```

Simulate genotype and lanc for given f1,f2
```{r}
#simulate genotypes for pop1 (200 loci and 1000 individuals), including non-causal variants
geno1 = replicate(nloci, rbinom(1000, 2, f1))
#simulate genotypes for pop2
geno2 = replicate(nloci, rbinom(1000, 2, f2))

```

Sample effect sizes for the 200 loci such that the genetic variance in the total population is 1.

```{r}
#proportion of loci that are causal
pcausal = 0.2 #1000 loci, 200 causal variants
ncausal = round(pcausal*nloci) # get number of causal variants
#choose effect sizes for causal loci, such that the total population variance is 1
bg = matrix(
  rnorm(ncausal, 0, sd= sqrt(vgt/(ncausal*2*fbar*(1-fbar)))), 
  nrow = ncausal, ncol = 1)
#standardize to remove stochasticity in realized values
bg = ((bg - mean(bg)) / sd(bg)) * sqrt(vgt/(ncausal*2*fbar*(1-fbar)))
#add 0s for non-causal loci
bg = sample(c(bg, rep(0, nloci - ncausal)))
```


```{r}
#calculate variance between populations
pbar = (f1 + f2)/2
vtotal = sum(2 * bg^2 * pbar * (1 - pbar))
vwithin = sum(2*(bg^2)*f1*(1 - f1) + 2*(bg^2)*f2*(1 - f2))/2
vbetween = vtotal - vwithin
```


Simulate local ancestry and genotypes for admixed population such that the admixture fraction is 0.5.

Local ancestry at any single locus is assmued to be Binomial with N = 2 and mean = 0.5, the admixture fraction. In other words, the probability of drawing ancestry from population 1 at any locus is 0.5. We assume loci are independent of each other (no LD).

```{r}

#simulate local ancestry in admixed population
#admixture fraction: 0.5
lanc = replicate(nloci, rbinom(10000, 2, theta))
mean(lanc[1,])
```

Now simulate genotypes at each locus given the ancestry at that locus. If the ancestry of an allele is from pop1, the probability that it is (say the 'A') allele is 0.8. The probability of it being the same allele if the ancestry is from pop2 is 0.3.

```{r }
#simulate genotypes
geno.admix = structure(sapply(lanc, function(x){
  if(x == 0){
    g1 = rbinom(1, 1, f1)
    g2 = rbinom(1, 1, f1)
  }
  if(x == 1){
    g1 = rbinom(1, 1, f1)
    g2 = rbinom(1, 1, f2)
  }
  if(x == 2){
    g1 = rbinom(1, 1, f2)
    g2 = rbinom(1, 1, f2)
  }
  return(g1 + g2)
}), dim = dim(lanc))

```

Simulate genetic values in the admixed population.

```{r}
#generate genetic value in admixed population
gvalue.admix = t(t(bg)%*%t(geno.admix))
sample_size=length(gvalue.admix)
# add environment noise to gvalue to create the pheno data
# in order to get vg/(vg+ve) = 0.8, simulate ve with var=0.25
phenotype = as.data.table(gvalue.admix)
colnames(phenotype)=c("gvalue")
phenotype$environment = rnorm(n = sample_size, mean = 0,
                        sd = sqrt(0.25))
# normalize gvalue
phenotype$gvalue_norm = (phenotype$gvalue - mean(phenotype$gvalue))/sd(phenotype$gvalue)

#add prs to each of the environmental effects
phenotype$pheno = phenotype$gvalue_norm + phenotype$environment

print(paste("correlation:",
            round(cor(phenotype$gvalue,
                      phenotype$pheno)^2,3)))
pheno<-data.frame(FID="ADM",IID=1:10000, phenovalue=phenotype$pheno)

```





Using linear model to estimate the effect size from simulated gvalue. 
Validate if the estimated effect size match the simulated effect size, and Validate if the estimated effect size of local ancestry match the converted simulated effect size 
 
```{r}
#estimate effect size for genotypes and local ancestry
est.b.g = matrix(NA, nrow = nloci, ncol = 1)
est.b.l = matrix(NA, nrow = nloci, ncol = 1)

# change gvalue to the pheno
lanc<-as.matrix(lanc)
geno.admix<-as.matrix(geno.admix)
for(i in 1:nloci){
  l1 = lm(phenotype$pheno ~ geno.admix[,i])
  s1 = summary(l1)$coefficients
  est.b.g[i,1] = s1[2,1]
  
  l2 = lm(phenotype$pheno ~ lanc[,i])
  s2 = summary(l2)$coefficients
  est.b.l[i,1] = s2[2,1]
}

```

Expected effect size of local ancestry (based on analytical derivation). Then, plot the expected vs estimated effect size of local ancestry.

```{r}
bl = bg * (f2 - f1)#
#estimated and expected effect size of genotype
plot(bg, est.b.g)
#estimate and expected effect size of ancestry
plot(bl, est.b.l)
```
The estimated effect size of ancestry is slightly noisier. Will have to work out later what the power of this association test is.


Now, calculate the expected and estimated (using effect sizes not GRM) values of genetic variance due to local ancestry. 


```{r}
#calculate expected genetic variance due to local ancestry
library(data.table)
#approach 1: using true genotype effect sizes
exp.vg.ancestry1 = 2*theta*(1 - theta)*sum(bg^2  * (f1 - f2)^2)
#approach 2: using true local ancestry effect sizes
exp.vg.ancestry2 = 2*theta*(1 - theta)*sum(bl^2)
#calculate observed genetic variance due to local ancestry
#approach 1: generate genetic values using estimated effect sizes and calculate variance in genetic values
gvalues.est.lanc = t(t(est.b.l) %*% t(lanc))
obs.vg.ancestry1 = var(gvalues.est.lanc)
#approach 2: square the effect sizes of ancestry and so on.
obs.vg.ancestry2 = sum(est.b.l^2 * 
      2*(apply(lanc, 2, mean)/2) * 
      ( 1- apply(lanc, 2, mean)/2))
data.table(expected = c(exp.vg.ancestry1, exp.vg.ancestry2), 
           observed = c(obs.vg.ancestry1, obs.vg.ancestry2))
```
```{r}
# convert admix lanc to tped 
# change lanc to certain allele
lanc<-as.data.table(lanc)
lanc[lanc == 0] <- "A,A"
lanc[lanc == 1] <- "A,T"
lanc[lanc == 2] <- "T,T"
# split them into 2 rows separately
lanc.allele<-data.frame(lapply(lanc, function(x) unlist(strsplit(as.character(x), ","))))

# transpose it and add map info: chrom, ID, cm (0), position
lanc.t<-t(lanc.allele)
map<-data.frame(chrom ="1",  cm=0, pos=1:nloci)
map$ID <- paste0(map$chrom, "_", map$pos, "_AT")

# reorder
map<-map[,c(1,4,2,3)]

# bind map and lanc together
l=cbind(map, lanc.t)

# output as tped file
write.table(l, file = paste0("admix", filename, ".tped"), quote=FALSE, sep = '\t' ,
            row.names = FALSE, col.names = FALSE)

# make tfam file
fam<-data.frame(FID="ADM",IID=1:10000,FID=0,MID=0,SEX=0,Pheno=-9 )
write.table(fam, file = paste0("admix", filename, ".tfam"), quote=FALSE, sep = '\t' ,
            row.names = FALSE, col.names = FALSE)
```




Convert genotype data to tped for gcta estimate. (espected h2=0.8)
```{r}
# change geno to certain allele
geno.admix<-as.data.table(geno.admix)
geno.admix[geno.admix == 0] <- "A,A"
geno.admix[geno.admix == 1] <- "A,T"
geno.admix[geno.admix == 2] <- "T,T"
# split them into 2 rows separately
geno.allele<-data.frame(lapply(geno.admix, function(x) unlist(strsplit(as.character(x), ","))))

# transpose it and add map info: chrom, ID, cm (0), position
geno.t<-t(geno.allele)
map<-data.frame(chrom ="1",  cm=0, pos=1:nloci)
map$ID <- paste0(map$chrom, "_", map$pos, "_AT")

# reorder
map<-map[,c(1,4,2,3)]

# bind map and lanc together
l=cbind(map, geno.t)

# output as tped file
write.table(l, file = paste0("admix_geno", filename, ".tped"), quote=FALSE, sep = '\t' ,
            row.names = FALSE, col.names = FALSE)

# make tfam file
fam<-data.frame(FID="ADM",IID=1:10000,FID=0,MID=0,SEX=0,Pheno=-9 )
write.table(fam, file = paste0("admix_geno", filename, ".tfam"), quote=FALSE, sep = '\t' ,
            row.names = FALSE, col.names = FALSE)


```


# reformat
```{r}
#!/usr/bin/env Rscript
#args = commandArgs(trailingOnly=TRUE)
# reformat the result to compare expected vs estimated vgb
# add geno gcta results in to the table
library(data.table)
#theta=args[1]
#vgb=args[2]
theta=0.5
vbg=0.3
simfile<-paste0("sim_prop",theta, "_vgb",vgb,"_expected.txt")
estfile<-paste0("admix_prop",theta, "_vgb",vgb, ".hsq")
genofile<-paste0("admix_geno_prop",theta, "_vgb",vgb, ".hsq")
  
sim<-fread(simfile)
est<-fread(estfile, fill=TRUE)
geno<-fread(genofile, fill=TRUE)
# function to reformat hsq file
reformat <- function(x){
  x<-x[complete.cases(x), ] # delete NA
  x<-as.data.frame(melt(x, id.vars = "Source"))
  x$Source <- paste(x$Source, x$variable, sep=".")
  x<-x[c("Source", "value")]
}
# reformat
geno<-reformat(geno)
est<-reformat(est)
#add extra for geno
geno$Source=paste0(geno$Source, ".geno")
# bind all together
colnames(sim)<-c("Source", "value")
est_sim<-rbind(sim, est, geno)
est_sim_t<-t(est_sim)
# output the reformat info
write.table(est_sim_t, file = paste0("simVSest_prop",theta, "_vgb",vgb,".txt"), 
            quote = F, col.names = F, row.names = F)

```

#plot
```{r}
library(data.table)
filename<-"simVSest_ALL.txt"
simVSest<-fread(filename)
library(ggplot2)

# calculate the upper and lower for errorbar 95% confidence interval
simVSest$adjustVG = simVSest$`V(G).Variance` /( 4 * simVSest$admixture * (1 - simVSest$admixture))
#Adjust SE or not?
simVSest$VG.SE=1.96*simVSest$`V(G).SE`/( 4 * simVSest$admixture * (1 - simVSest$admixture))
#simVSest$VG.SE=1.96*simVSest$`V(G).SE`

simVSest$adjustVG.lower=simVSest$adjustVG-simVSest$VG.SE 
simVSest$adjustVG.upper=simVSest$adjustVG+simVSest$VG.SE

# plot to compare vgb
png(file=paste0("simVSest_ALL_vgb_addenv.png"), width=12, height=6, units="in", res=300)
ggplot(simVSest, aes(x=admixture, y=adjustVG)) +
#ggplot(simVSest, aes(x=admixture, y=`V(G).Variance`)) +
  geom_errorbar(width=.005,aes(ymin = adjustVG.lower, ymax = adjustVG.upper)) +
  geom_point(shape=21, size=3, fill="white") +
  #add facet for different admixture
  facet_grid(cols = vars(vgb)) +
  #facet_grid(vars(vgb), vars(admixture))+
  ggtitle("Expected VS Estimated Between Population Genetic Variance Under different Admixture Proportion (CI=95%) Ve=0.2" ) + 
  xlab("Admixture Proportion") + ylab("Estimated Genetic Variance (Adjusted)") +
  #geom_abline(slope=0, intercept = vbetween, size = .5, color = "red") + 
  geom_point(aes(x=admixture, y=vbetween), colour="red") +
  #ylim(0,1.2) + 
  theme_bw()
dev.off()

# plot to compare vgb unadjusted
png(file=paste0("simVSest_ALL_vgb_addenv_notadjusted.png"), width=12, height=6, units="in", res=300)
ggplot(simVSest, aes(x=admixture, y=`V(G).Variance`)) +
  #geom_errorbar(width=.005,aes(ymin = adjustVG.lower, ymax = adjustVG.upper)) +
  geom_point(shape=21, size=3, fill="white") +
  #add facet for different admixture
  facet_grid(cols = vars(vgb)) +
  #facet_grid(vars(vgb), vars(admixture))+
  ggtitle("Expected VS Estimated Between Population Genetic Variance Under different Admixture Proportion (CI=95%) Ve=0.2" ) + 
  xlab("Admixture Proportion") + ylab("Estimated Genetic Variance (Not Adjusted)") +
  #geom_abline(slope=0, intercept = vbetween, size = .5, color = "red") + 
  geom_point(aes(x=admixture, y=vbetween), colour="red") +
  #ylim(0,1.2) + 
  theme_bw()
dev.off()

# plot to compare ve vs vwithin
png(file=paste0("simVSest_ALL_vwithin_addenv.png"), width=12, height=6, units="in", res=300)
ggplot(simVSest, aes(x=admixture, y=`V(e).Variance`-venv)) +
  #geom_errorbar(width=.005,aes(ymin = adjustVG.lower, ymax = adjustVG.upper)) +
  geom_point(shape=21, size=3, fill="white") +
  #add facet for different admixture
  facet_grid(cols = vars(vgb)) +
  #facet_grid(vars(vgb), vars(admixture))+
  ggtitle("Expected VS Estimated Within Population Genetic Variance Under different Admixture Proportion (CI=95%) Ve=0.2" ) + 
  xlab("Admixture Proportion") + ylab("Estimated Genetic Variance (Adjusted)") +
  #geom_abline(slope=0, intercept = vbetween, size = .5, color = "red") + 
  geom_point(aes(x=admixture, y=vwithin), colour="red") +
  #ylim(0,1.2) + 
  theme_bw()
dev.off()

# plot to compare vg/vp vs vgeno
simVSest$genoh2.lower=simVSest$`V(G)/Vp.Variance.geno` - simVSest$`V(G)/Vp.SE.geno`
simVSest$genoh2.upper=simVSest$`V(G)/Vp.Variance.geno`+simVSest$`V(G)/Vp.SE.geno`

png(file=paste0("simVSest_ALL_vgeno_addenv.png"), width=12, height=6, units="in", res=300)
ggplot(simVSest, aes(x=admixture, y=`V(G)/Vp.Variance.geno`)) +
  geom_errorbar(width=.005,aes(ymin = genoh2.lower, ymax = genoh2.upper)) +
  geom_point(shape=21, size=3, fill="white") +
  #add facet for different admixture
  facet_grid(cols = vars(vgb)) +
  #facet_grid(vars(vgb), vars(admixture))+
  ggtitle("Expected VS Estimated Genotype Heritability Under different Admixture Proportion (CI=95%) Ve=0.2" ) + 
  xlab("Admixture Proportion") + ylab("Estimated Genotype Heritability") +
  #geom_abline(slope=0, intercept = vbetween, size = .5, color = "red") + 
  geom_point(aes(x=admixture, y=vgeno), colour="red") +
  ylim(0,1) + 
  theme_bw()
dev.off()


```


